From 52efe222e9df5a632c9104db9f21ac59986b86a2 Mon Sep 17 00:00:00 2001
From: Abhishek Chauhan <quic_abchauha@quicinc.com>
Date: Mon, 24 Nov 2025 10:53:52 -0800
Subject: [PATCH 13/16] PENDING: net: PCS: Check for AN status before posting
 link up

Currently after serdes reset we check for just link status
bit in case of 10GbaseR and 5GbaseR cases which does not
support Auto-negotiation.

In case of USXGMII, Auto-negotiation is supported. We need to check
for AN status before we post link as up to the MAC.

Change-Id: I690c99c60821166177fa1cff1916ac3ceb4a8f0e
Signed-off-by: Abhishek Chauhan <quic_abchauha@quicinc.com>
(cherry picked from commit 17b30a37b6a76fe70d450353a4f820a9216af261)
---
 drivers/net/pcs/pcs-xpcs-qcom.c | 27 ++++++++++++++-------------
 1 file changed, 14 insertions(+), 13 deletions(-)

diff --git a/drivers/net/pcs/pcs-xpcs-qcom.c b/drivers/net/pcs/pcs-xpcs-qcom.c
index 64182274d038..6e54f476addf 100644
--- a/drivers/net/pcs/pcs-xpcs-qcom.c
+++ b/drivers/net/pcs/pcs-xpcs-qcom.c
@@ -484,6 +484,20 @@ static int qcom_xpcs_get_link_status(struct dw_xpcs_qcom *qxpcs,
 	if (!qxpcs->intr_en) {
 		switch (qxpcs->phy_interface) {
 		case PHY_INTERFACE_MODE_USXGMII:
+			if (qxpcs->needs_aneg) {
+				ret = qcom_xpcs_poll_bit_set(qxpcs,
+							     DW_VR_MII_AN_INTR_STS,
+							     DW_VR_MII_ANCMPLT_INTR);
+
+				if (ret < 0)
+					goto recover;
+				else
+					XPCSDBG("XPCS AN completed\n");
+				ret = qcom_xpcs_read(qxpcs, DW_VR_MII_AN_INTR_STS);
+				/* Clear the IOC status */
+				ret &= ~DW_VR_MII_ANCMPLT_INTR;
+				qcom_xpcs_write(qxpcs, DW_VR_MII_AN_INTR_STS, ret);
+			}
 			/* Check for Link status */
 			ret = qcom_xpcs_poll_bit_set(qxpcs,
 						     DW_SR_MII_MMD_STS, DW_SR_MII_STS_LINK_STS);
@@ -499,19 +513,6 @@ static int qcom_xpcs_get_link_status(struct dw_xpcs_qcom *qxpcs,
 				XPCSDBG("Link is down try to recover\n");
 				goto recover;
 			}
-			/* Check for block lock status */
-			ret = qcom_xpcs_poll_bit_set(qxpcs,
-						     DW_SR_MII_PCS_KR_STS2, DW_LAT_BL);
-			if (ret < 0) {
-				XPCSDBG("DW_LAT_BL goto recover\n");
-				goto recover;
-			}
-			/* Check for block error  status */
-			ret = qcom_xpcs_poll_reset(qxpcs, DW_SR_MII_PCS_KR_STS2, DW_ERR_BLK);
-			if (ret < 0) {
-				XPCSDBG("DW_ERR_BLK goto recover\n");
-				goto recover;
-			}
 		}
 		state->link = true;
 		return 0;
